src/pubsub/README.md

```diff
@@ -0,0 +1,218 @@
+
+All the remaining functions **MUST NOT** be overwritten.
+
+The following example aims to show how to create your pubsub implementation extending this base protocol. The pubsub implementation will handle the subscriptions logic.
+
+```JavaScript
+const Pubsub = require('libp2p-pubsub')
+
+class PubsubImplementation extends Pubsub {
+  constructor({ peerId, registrar, ...options })
+    super({
+      debugName: 'libp2p:pubsub',
+      multicodecs: '/pubsub-implementation/1.0.0',
+      peerId: peerId,
+      registrar: registrar,
+      signMessages: options.signMessages,
+      strictSigning: options.strictSigning
+    })
+  }
+
+  _processMessages(idB58Str, conn, peer) {
+    // Required to be implemented by the subclass
+    // Process each message accordingly
+  }
+
+  publish() {
+    // Required to be implemented by the subclass
+  }
+
+  subscribe() {
+    // Required to be implemented by the subclass
+  }
+
+  unsubscribe() {
+    // Required to be implemented by the subclass
+  }
+
+  getTopics() {
+    // Required to be implemented by the subclass
+  }
+}
+```
+
+## API
+
+The interface aims to specify a common interface that all pubsub router implementation should follow.
+
+### Start
+
+Starts the pubsub subsystem. The protocol will be registered to `libp2p`, which will result in pubsub being notified when peers who support the protocol connect/disconnect to `libp2p`.
+
+#### `pubsub.start()`
+
+##### Returns
+
+| Type | Description |
+|------|-------------|
+| `Promise<void>` | resolves once pubsub starts |
+
+### Stop
+
+Stops the pubsub subsystem. The protocol will be unregistered from `libp2p`, which will remove all listeners for the protocol and the established connections will be closed.
+
+#### `pubsub.stop()`
+
+##### Returns
+
+| Type | Description |
+|------|-------------|
+| `Promise<void>` | resolves once pubsub stops |
+
+### Publish
+
+Publish data message to pubsub topics.
+
+#### `pubsub.publish(topics, message)`
+
+##### Parameters
+
+| Name | Type | Description |
+|------|------|-------------|
+| topics | `Array<string>|string` | set of pubsub topics |
+| message | `Uint8Array` | message to publish |
+
+##### Returns
+
+| Type | Description |
+|------|-------------|
+| `Promise<void>` | resolves once the message is published to the network |
+
+### Subscribe
+
+Subscribe to the given topic(s).
+
+#### `pubsub.subscribe(topics, [handler])`
+
+##### Parameters
+
+| Name | Type | Description |
+|------|------|-------------|
+| topics | `Array<string>|string` | set of pubsub topics |
+| [handler] | `function (msg)` | handler for messages received in the given topics |
```

Comment:
I think the interface should only allow subscription/unsubscription of a single topic at a time.

Attaching a handler to multiple topics at once feels like an anti-pattern, or at least a less primary interface, in the same way we were previously attaching multiple topics to multiple messages at once in `publish`.

I think this will simplify some of the underlying implementations, too. The body of our subscription code looks roughly like: `for (const topic of topics) { /* doSubscribe(topic) */ }`. We would be able to remove this outer loop and better support extended behavior.

Also, go-libp2p-pubsub's interface for subscribing is for a single topic at a time, for what its worth.

Reply:
I agree that the API would look better with one topic. Taking into account that we are doing several breaking changes in pubsub, I think we should also change this.

The libp2p API doc also says a [single](https://github.com/libp2p/js-libp2p/blob/master/doc/API.md#pubsubsubscribe) topic, so we were supporting it below in the stack but we were not mentioning it. However, this should be pointed in the migration guide

Reply:
What about publish? Should we also just accept one topic per API call? That's how we have it in libp2p main docs as well

Reply:
I will update it